cc/mcts_player.cc

```diff
@@ -138,46 +128,84 @@ void MctsPlayer::TreeSearch(int batch_size) {
       leaf->IncorporateEndGameResult(value, root_);
     } else {
       leaf->AddVirtualLoss(root_);
-      leaves.push_back(leaf);
-      if (static_cast<int>(leaves.size()) == batch_size) {
+      leaves_.push_back(leaf);
+      if (static_cast<int>(leaves_.size()) == batch_size) {
         break;
       }
     }
   }
 
-  if (!leaves.empty()) {
-    // TODO(tommadams): Avoid creating these vectors each time.
-    std::vector<const DualNet::BoardFeatures*> features;
-    features.reserve(leaves.size());
-    for (auto* leaf : leaves) {
-      features.push_back(&leaf->features);
+  if (!leaves_.empty()) {
+    features_.clear();
+    features_.reserve(leaves_.size());
+    for (auto* leaf : leaves_) {
+      features_.push_back(&leaf->features);
     }
 
-    std::vector<DualNet::Output> outputs;
-    outputs.resize(leaves.size());
+    outputs_.resize(leaves_.size());
```

Comment:
can you get away with `reserve` as the values don't need to be initialized?
This might be premature optimization but I think this is on the critical/hot path

Reply:
no, you can't legally access the elements of a vector if you only call reserve. This isn't an issue because in practice outputs_ is almost always the same size.